src/compiler/scala/tools/nsc/transform/patmat/PatternMatching.scala

```diff
@@ -61,7 +61,10 @@ trait PatternMatching extends Transform
         // setType origTp intended for CPS -- TODO: is it necessary?
         val translated = translator.translateMatch(treeCopy.Match(tree, transform(sel), transformTrees(cases).asInstanceOf[List[CaseDef]]))
         try {
-          localTyper.typed(translated) setType origTp
+          // Keep 2.12 behaviour of using wildcard expected type, recomputing the LUB, then throwing it away for the continuations plugins
+          // but for the rest of us pass in top as the expected type to avoid waste.
+          val pt = if (origTp <:< definitions.AnyTpe) definitions.AnyTpe else WildcardType
+          localTyper.typed(translated, definitions.AnyTpe) setType origTp
```

Comment:
I guess no one cares about continuations, but did you intend to use `pt` here? "-Xlint would have told you that". -- Paul.

Reply:
Thanks for fixing this down the line. It wasn't even the only bug in this patch, #7016 fixes another.